File: solve_SDP_multi.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 21 02:15:59 2020

@author: paul_
"""


import torch
import matlab.engine
import numpy as np
import json


eng = matlab.engine.start_matlab()


def build_T_multi(weights, biases, net_dims):
    x = matlab.int64([net_dims])

    parameters = {}
    weights_dict = {}
    biases_dict = {}
    for i in range(len(weights)):
        parameters.update({
            'W{:d}'.format(i): matlab.double(np.array(weights, dtype=np.object)[i].tolist()),
            })
        parameters.update({
            'b{:d}'.format(i): matlab.double(np.array(biases, dtype=np.object)[i].tolist()),
            })
        weights_dict.update({
            'W{:d}'.format(i): np.array(weights, dtype=np.object)[i].tolist() })
        biases_dict.update({
            'b{:d}'.format(i): np.array(biases, dtype=np.object)[i].tolist() })

    with open('weights.json', 'w') as f:  # writing JSON object
        json.dump(weights_dict, f)
    with open('biases.json', 'w') as f:  # writing JSON object
        json.dump(biases_dict, f)
    Lip = eng.calculate_Lipschitz_multi(parameters, x)
    return Lip


def solve_SDP_multi(parameters, T, net_dims, rho, mu, ind_Lip, L_des):
    x = matlab.int64([net_dims])
    sdp = {
        'T': matlab.double([T]),
        'rho': matlab.double([rho]),
        'mu': matlab.double([mu]),
        'ind_Lip': matlab.int64([ind_Lip]),
        'L_des': matlab.double([L_des])
    }

    parameters = eng.solve_sdp_multi(parameters, x, sdp)

    return parameters


def initialize_parameters(weights, biases):
    parameters = {}
    for i in range(len(weights)):
        parameters.update({
            'W{:d}'.format(i): matlab.double(np.array(weights, dtype=np.object)[i].tolist()),
            })
        parameters.update({
            'W{:d}_bar'.format(i): matlab.double(np.array(weights, dtype=np.object)[i].tolist()),
            })
        parameters.update({
            'b{:d}'.format(i): matlab.double(np.array(biases, dtype=np.object)[i].tolist()),
            })
        parameters.update({
            'Y{:d}'.format(i): matlab.double(torch.zeros(weights[i].shape).tolist()),
            })
        parameters.update({
            'Yb{:d}'.format(i): matlab.double(torch.zeros(biases[i].shape).tolist()),
            })
    return parameters

File: solve_sdp_multi.m
function parameters=solve_sdp_multi(parameters, x, sdp)
%[parameters] = solve_sdp(n_x, n_h, n_y, lambd, mu, Lip, ind_Lip, parameters)
%test

addpath(genpath('..\YALMIP-master'))
addpath(genpath('C:\Program Files\Mosek\9.2'))

%net_dim = x;
% net_dim = [2, 10, 10, 3];

fname = 'weights.json';
weights = jsondecode(fileread(fname));
fn = fieldnames(weights);
for k=1:numel(fn)
    W{k}= weights.(fn{k});
    net_dim(k)=size(W{k},2); % PP: get dimensions from weights
end
net_dim(k+1)=size(W{k},1); % PP: see above
fname = 'biases.json';
biases = jsondecode(fileread(fname));
fn = fieldnames(biases);
for k=1:numel(fn)
    b{k} = biases.(fn{k});
end


l = length(net_dim)-1;
eps = 10^(-9);

lambd = double(sdp.rho);
mu = double(sdp.mu);
ind_Lip = int64(sdp.ind_Lip); % 1 Lipschitz regularizaion, 2 Enforcing Lipschitz bounds
Lip = double(sdp.L_des);
T = squeeze(sdp.T);
t = diag(T);
%for k = 2:l
%    Ts{k} = diag(t(sum(net_dim(2:k-1))+1:sum(net_dim(2:k))));
%end


for k=1:l
    W_bar{k} = sdpvar(net_dim(k+1), net_dim(k));
    Y{k} = parameters.(['Y' num2str(k-1)]);
end

A = [blkdiag(W_bar{1:l-1}), zeros(sum(net_dim(2:l)), sum(net_dim(l:l+1)))];
B = [zeros(sum(net_dim(2:l)), net_dim(1)), eye(sum(net_dim(2:l))), zeros(sum(net_dim(2:l)), net_dim(l+1))];
C = [zeros(net_dim(length(net_dim)), sum(net_dim(1:l-1))), W{l}];

W_diff = 0;
sumOfTraces = 0;
for k=1:l
    W_diff = W_diff + norm(W{k}-W_bar{k}, 'fro')^2 * (lambd/2); % PP: added ^2
    sumOfTraces = sumOfTraces + trace(Y{k}'*(W{k}-W_bar{k})); % PP: added trace()
end


if ind_Lip==1
    rho = sdpvar(1);
    Q_untenrechts = [zeros(net_dim(l)), W_bar{l}'; W_bar{l}, -eye(net_dim(l+1))]; % PP: changed W to W_bar
    Q = blkdiag(-rho*eye(net_dim(1)),zeros(sum(net_dim(2:l-1))), Q_untenrechts);
    %X = blkdiag(zeros(net_dim(1)), Ts{2:l}, zeros(net_dim(l+1)));
    %for k = 2:l
    %    X(sum(net_dim(1:k-2))+1:sum(net_dim(1:k-1)), sum(net_dim(1:k-1))+1:sum(net_dim(1:k))) = W_bar{k-1}'*Ts{k}; % PP: changed W to W_bar
    %    X(sum(net_dim(1:k-1))+1:sum(net_dim(1:k)), sum(net_dim(1:k-2))+1:sum(net_dim(1:k-1))) = Ts{k}*W_bar{k-1}; % PP: changed W to W_bar
    %end
    X2=[A;B]'*[0*T, T; T, -2*T]*[A;B]; % PP: Reformulation of X, check if yours does the same, seemed like factor -2 was missing, adapt in ind_Lip=2

    M = X2 + Q;
    F1 = (M <= -eps*eye(sum(net_dim(1:l+1))));
    optimize(F1, mu*rho + W_diff + sumOfTraces);
elseif ind_Lip==2
    rho = Lip^2;
    Q_untenrechts = [zeros(net_dim(l)), W_bar{l}'; W_bar{l}, -eye(net_dim(l+1))];
    Q = blkdiag(-rho*eye(net_dim(1)),zeros(sum(net_dim(2:l-1))), Q_untenrechts);
%    X = blkdiag(zeros(net_dim(1)), Ts{2:l}, zeros(net_dim(l+1)));
%    for k = 2:l
%        X(sum(net_dim(1:k-2))+1:sum(net_dim(1:k-1)), sum(net_dim(1:k-1))+1:sum(net_dim(1:k))) = W_bar{k-1}'*Ts{k};
%        X(sum(net_dim(1:k-1))+1:sum(net_dim(1:k)), sum(net_dim(1:k-2))+1:sum(net_dim(1:k-1))) = Ts{k}*W_bar{k-1};
%    end
    X2 = [A;B]'*[0*T,T;T-2*T]*[A;B]; %M = X + Q;
    M = X2 + Q;
    F1 = (M <= -eps*eye(sum(net_dim(1:l+1))));
    optimize(F1, mu*rho + W_diff + sumOfTraces);
end

Lipschitz = sqrt(value(rho));

for k=1:l
    W_bar{k} = value(W_bar{k});
    Y{k} = Y{k} + lambd * (W{k}-W_bar{k}); % dual update step
    parameters.(['W' num2str(k-1) '_bar']) = value(W_bar{k}); % PP: changes
    parameters.(['Y' num2str(k-1)]) = Y{k};
end

parameters.Lipschitz=Lipschitz;

